From ee30fe0b9ac17ace00d876fa85dbcbf18e08c56e Mon Sep 17 00:00:00 2001
From: Roger Siddons <dizygotheca@ntlworld.com>
Date: Thu, 30 Oct 2014 19:10:23 +0000
Subject: [PATCH 10/11] New Image Gallery: Fix rotate/flip

Adds storage group support for transforms and moves cache management to
client.

Signed-off-by: Paul Harrison <pharrison@mythtv.org>
---
 mythtv/libs/libmythmetadata/imagethumbgenthread.cpp |    4 ----
 mythtv/programs/mythbackend/services/image.cpp      |   12 ++++++++----
 mythtv/programs/mythfrontend/galleryview.cpp        |   19 ++++++++++++-------
 mythtv/programs/mythfrontend/galleryview.h          |    2 +-
 mythtv/programs/mythfrontend/gallerywidget.cpp      |    5 ++++-
 5 files changed, 25 insertions(+), 17 deletions(-)

diff --git a/mythtv/libs/libmythmetadata/imagethumbgenthread.cpp b/mythtv/libs/libmythmetadata/imagethumbgenthread.cpp
index 6166f6f..a5cb409 100644
--- a/mythtv/libs/libmythmetadata/imagethumbgenthread.cpp
+++ b/mythtv/libs/libmythmetadata/imagethumbgenthread.cpp
@@ -270,12 +270,8 @@ void ImageThumbGenThread::RecreateThumbnail(ImageMetadata *im)
         return;
 
     if (QFile::remove(im->m_thumbFileNameList->at(0)))
-    {
-        GetMythUI()->RemoveFromCacheByFile(
-                    im->m_thumbFileNameList->at(0));
 
         AddToThumbnailList(im);
-    }
 }
 
 
diff --git a/mythtv/programs/mythbackend/services/image.cpp b/mythtv/programs/mythbackend/services/image.cpp
index 7ef7d9c..c7578c0 100644
--- a/mythtv/programs/mythbackend/services/image.cpp
+++ b/mythtv/programs/mythbackend/services/image.cpp
@@ -51,17 +51,21 @@ bool Image::SetImageInfo( int id, const QString &tag,
     ImageUtils *iu = ImageUtils::getInstance();
     iu->LoadFileFromDB(im, id);
 
-    if (im->m_fileName.isEmpty())
+    QString sgName = IMAGE_STORAGE_GROUP;
+    StorageGroup sg = StorageGroup(sgName, gCoreContext->GetHostName());
+    QString imageFileName = sg.FindFile(im->m_fileName);
+
+    if (imageFileName.isEmpty())
     {
-        LOG(VB_GENERAL, LOG_ERR, "SetImageInfo - File not found in DB.");
+        LOG(VB_GENERAL, LOG_ERR,
+            QString("SetImageInfo - Image %1 not found in DB.").arg(id));
         delete im;
         return false;
     }
 
     // We got the file name from the ID, so use this method
     // which does the same but just on a filename basis.
-    bool ok;
-    ok = SetImageInfoByFileName( im->m_fileName, tag, value );
+    bool ok = SetImageInfoByFileName(imageFileName, tag, value);
 
     delete im;
     return ok;
diff --git a/mythtv/programs/mythfrontend/galleryview.cpp b/mythtv/programs/mythfrontend/galleryview.cpp
index d891d3a..c77acd0 100644
--- a/mythtv/programs/mythfrontend/galleryview.cpp
+++ b/mythtv/programs/mythfrontend/galleryview.cpp
@@ -9,7 +9,7 @@
 #include "galleryconfig.h"
 #include "gallerytypedefs.h"
 #include "imagescan.h"
-
+#include "mythuihelper.h"
 
 
 /** \fn     GalleryView::GalleryView(MythScreenStack *, const char *)
@@ -439,7 +439,7 @@ void GalleryView::UpdateImageList()
  *  \param  item The item that shall be updated
  *  \return void
  */
-void GalleryView::UpdateImageItem(MythUIButtonListItem *item)
+void GalleryView::UpdateImageItem(MythUIButtonListItem *item, bool recreateThumb)
 {
     if (!item)
         return;
@@ -514,8 +514,13 @@ void GalleryView::UpdateImageItem(MythUIButtonListItem *item)
     for (int i = 0; i < im->m_thumbFileIdList.size(); ++i)
         m_imageMap.insert(im->m_thumbFileIdList.at(i), item);
 
+    if (recreateThumb)
+        // remove cache image to force a reload
+        GetMythUI()->RemoveFromCacheByFile(
+                    im->m_thumbFileNameList->at(0));
+
     // set the thumbnail image
-    UpdateThumbnail(item, im);
+    UpdateThumbnail(item, im, recreateThumb);
 }
 
 
@@ -1115,7 +1120,7 @@ void GalleryView::FileRotateCW()
         return;
 
     m_galleryViewHelper->SetFileOrientation(kFileRotateCW);
-    UpdateImageItem(item);
+    UpdateImageItem(item, true);
 }
 
 
@@ -1132,7 +1137,7 @@ void GalleryView::FileRotateCCW()
         return;
 
     m_galleryViewHelper->SetFileOrientation(kFileRotateCCW);
-    UpdateImageItem(item);
+    UpdateImageItem(item, true);
 }
 
 
@@ -1149,7 +1154,7 @@ void GalleryView::FileFlipHorizontal()
         return;
 
     m_galleryViewHelper->SetFileOrientation(kFileFlipHorizontal);
-    UpdateImageItem(item);
+    UpdateImageItem(item, true);
 }
 
 
@@ -1166,7 +1171,7 @@ void GalleryView::FileFlipVertical()
         return;
 
     m_galleryViewHelper->SetFileOrientation(kFileFlipVertical);
-    UpdateImageItem(item);
+    UpdateImageItem(item, true);
 }
 
 
diff --git a/mythtv/programs/mythfrontend/galleryview.h b/mythtv/programs/mythfrontend/galleryview.h
index f56a418..45ea957 100644
--- a/mythtv/programs/mythfrontend/galleryview.h
+++ b/mythtv/programs/mythfrontend/galleryview.h
@@ -44,7 +44,7 @@ private:
 
 private slots:
     void ItemSelected(MythUIButtonListItem *);
-    void UpdateImageItem(MythUIButtonListItem *);
+    void UpdateImageItem(MythUIButtonListItem *, bool = false);
     void UpdateText(MythUIButtonListItem *);
     void UpdateThumbnail(MythUIButtonListItem *,
                          ImageMetadata *,
diff --git a/mythtv/programs/mythfrontend/gallerywidget.cpp b/mythtv/programs/mythfrontend/gallerywidget.cpp
index d22b29c4..603848f 100644
--- a/mythtv/programs/mythfrontend/gallerywidget.cpp
+++ b/mythtv/programs/mythfrontend/gallerywidget.cpp
@@ -2,12 +2,12 @@
 #include "gallerywidget.h"
 
 // Qt headers
-#include <QByteArray>
 #include <QXmlStreamReader>
 
 // MythTV headers
 #include "mythcontext.h"
 #include "mythmainwindow.h"
+#include "mythuihelper.h"
 
 ImageLoadingThread::ImageLoadingThread() :
     m_image(NULL),
@@ -538,6 +538,9 @@ void GalleryWidget::LoadFile()
 
         QString url = CreateImageUrl(fileName);
 
+        // remove cache image to force a reload
+        GetMythUI()->RemoveFromCacheByFile(fileName);
+
         // This thread will loads the image so the UI is not blocked.
         m_ilt->setImage(m_fileList->at(m_index),
                         m_fileDataList->at(m_index), url);
-- 
1.7.10.2

